etna/datasets/utils.py

```diff
@@ -484,3 +486,147 @@ def timestamp_range(
         return pd.Index(np.arange(start, start + periods))
     else:
         return pd.date_range(start=start, end=end, periods=periods, freq=freq)
+
+
+def infer_alignment(df: pd.DataFrame) -> Union[Dict[str, pd.Timestamp], Dict[str, int]]:
+    """Inference alignment of a given dataframe.
+
+    Alignment tells us which timestamps for each segment should be considered to have the same integer timestamp after
+    alignment transformation.
+
+    For long dataframe the alignment is determined by the last timestamp for each segment.
+    Last timestamp is taken without checking is 'target' value missing or not.
+
+    Parameters
+    ----------
+    df:
+        Dataframe in a long format.
+
+    Returns
+    -------
+    :
+        Dictionary with mapping segment -> timestamp.
+    """
+    return df.groupby(by=["segment"]).agg({"timestamp": "max"})["timestamp"].to_dict()
+
+
+def apply_alignment(
+    df: pd.DataFrame,
+    alignment: Union[Dict[str, pd.Timestamp], Dict[str, int]],
+    original_timestamp_name: Optional[str] = None,
+):
+    """Apply given alignment to a dataframe.
+
+    Applying alignment creates a new dataframe in which we have a new 'timestamp' column
+    with sequential integer timestamps.
+
+    For each segment we sort timestamps and assign them sequential integer values (with step 1)
+    in a way that timestamp in ``alignment`` gets value 0.
+
+    Parameters
+    ----------
+    df:
+        Dataframe in a long format.
+    alignment:
+        Alignment to apply.
+    original_timestamp_name:
+        Name for a column to save the original timestamps. If ``None`` original timestamps won't be saved.
+
+    Returns
+    -------
+    :
+        Aligned dataframe in a long format.
+
+    Raises
+    ------
+    ValueError:
+        There is a segment in ``df`` which isn't present in ``alignment``.
+    ValueError:
+        There is a segment which doesn't have a timestamp that is present in ``alignment``.
+    """
+    df_list = []
+    for segment in df["segment"].unique():
+        if segment not in alignment:
+            raise ValueError(f"The segment '{segment}' isn't present in alignment!")
+
+        cur_df = df[df["segment"] == segment].sort_values(by="timestamp")
+        reference_timestamp = alignment[segment]
+
+        if reference_timestamp not in cur_df["timestamp"].values:
+            raise ValueError(
+                f"The segment '{segment}' doesn't contain timestamp '{reference_timestamp}' from alignment!"
+            )
+
+        reference_timestamp_index = pd.Index(cur_df["timestamp"]).get_loc(reference_timestamp)
+        new_timestamp = np.arange(len(cur_df)) - reference_timestamp_index
+
+        if original_timestamp_name is not None:
+            cur_df.rename(columns={"timestamp": original_timestamp_name}, inplace=True)
+
+        cur_df["timestamp"] = new_timestamp
+        df_list.append(cur_df)
+
+    result = pd.concat(df_list)
+    return result
+
+
+def make_timestamp_df(
```

Comment:
Maybe we should name it something like "make_timestamp_df_from_alighnment" to show that it is a specific function?

Reply:
I'll think about it.